--- t1lib-5.1.2.orig/configure  2015-06-28 23:12:59.394297125 +0200
+++ t1lib-5.1.2/configure 2015-06-28 23:15:44.066290147 +0200
@@ -24758,11 +24758,7 @@
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   if test "$cross_compiling" = yes; then
-  { { echo "$as_me:$LINENO: error: cannot run test program while cross compiling
-See \`config.log' for more details." >&5
-echo "$as_me: error: cannot run test program while cross compiling
-See \`config.log' for more details." >&2;}
-   { (exit 1); exit 1; }; }
+  ac_64bit_type="<none>"
 else
   cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
