===================================================================
RCS file: /cvs/src/src/newlib/libc/machine/mips/memcpy.S,v
retrieving revision 1.1
retrieving revision 1.2
diff -u -r1.1 -r1.2
--- src/newlib/libc/machine/mips/memcpy.S	2012/12/11 21:37:59	1.1
+++ src/newlib/libc/machine/mips/memcpy.S	2013/01/07 21:29:39	1.2
@@ -56,7 +56,7 @@
 #endif
 #endif

-#if (_MIPS_SIM == _ABI64) || (_MIPS_SIM == _ABIN32)
+#if defined(_MIPS_SIM) && ((_MIPS_SIM == _ABI64) || (_MIPS_SIM == _ABIN32))
 #ifndef DISABLE_DOUBLE
 #define USE_DOUBLE
 #endif
@@ -203,6 +203,6 @@
 #define REG1 t1
 #define REG2 t2
 #define REG3 t3
-#if _MIPS_SIM == _ABIO32
+#if defined(_MIPS_SIM) && (_MIPS_SIM == _ABIO32 || _MIPS_SIM == _ABIO64)
 #  define REG4 t4
 #  define REG5 t5
 #  define REG6 t6
